Adding a blog on tips and tricks for handling multiple Kubernetes clusters with Kubectl (#10)
SUSE-Developer-Community/web-content: content/handling-multiple-k8sclusters.md

* `kubectl config --kubeconfig=config-demo set-context` to define or change contexts

Once you have set up your clusters, users and contexts, you can easily switch between contexts using the `kubectl config use-context` command. The currently active context can be checked with `kubectl config current-context`. 

A more detailed description of all this can be found at [https://kubernetes.io/docs/tasks/access-application-cluster/configure-access-multiple-clusters/]. 

Theoretically, this could be the end of the article, if it wasn't for the fact that once the number of clusters you interact with grows beyond low single-digit numbers, keeping all information in one config file becomes, let's say, unpleasant. 

## Multiple config files

A better way (in my humble opinion) is to keep information about different clusters in different config files. There's basically two ways to handle multiple config files: the simple way and the clever way. The simple way is you manually specify the config you want to use each time you call kubectl via the `--kubeconfig` flag. The clever way is to concatenate multiple kubeconfigs into your `KUBECONFIG` environment variable like 

>     export KUBECONFIG=$HOME/.kube/config:second-config-file:third-config-file. 

Reviewer: Have you tested this? I couldn't get it to work the last time I tried but maybe I was doing something wrong...

Author: I'm running this on my machine and it works fine.

* Files specified via --kubeconfig always win, no matter what's in your KUBECONFIG 
* The content of all configs specified in $KUBECONFIG is merged. You can use `kubectl config view --merge` to check what this leads to on your system. 
* The first file to set a value or map key wins
* Never change a value or map key. Even if a later occurrence has non-conflicting values, discard that info
* The first context found in the chain wins - if you have a context with the same name in a later file, kubectl will never find it. 

In case you ended up with multiple configs in one file, and you want to separate them, you can extract all information for one specific context with

    kubectl config view --minify --flatten --context=my_context > myconfig.yaml

## Auto-populate $KUBECONFIG

If the set of clusters you interact with changes often, one elegant way to deal with that is to populate your KUBECONFIG variable automatically. There's many ways to do this, but here's what I'm using. I have this line in my environment:

    set -xg KUBECONFIG $HOME/.kube/config:(ls -m $HOME/.kube/*.yaml | tr ',' ':' | tr -d "[:blank:]") 

>  Note that I'm using fish as my default shell, hence I'm using fish's native `set` command with the `-x` flag instead of `export`. In a normal bash or zsh environment you would use `export KUBECONFIG=...` for the same purpose. 

Reviewer: There might be room for a whole "how to use fish with kubernetes everything"

Author: Is there really that much more to say?

Reviewer: Probably not. I wrote this before seeing your comments later on. One problem that I run into commonly is that people write installation instructions for their tools that assume bash and use substitutions in a way that I can't just copy and paste

Reviewer: Which is out of scope for this article anyways

This picks up `~/.kube/config as well as all files with a .yaml file extension from the ~/.kube directory, and stores them as colon separated list in my KUBECONFIG variable. So if I add a new cluster, I just have to put the config there and re-source my environment. 

## Useful tools and tricks

So far we've obviously only covered Kubernetes' built-in functionality. The are a couple of little helpers and additional tricks that are worth knowing about. Disclaimer: this is not intended to be a complete list - there's tons of other stuff out there. Just sharing a few of my favorites. 

The first tool I want to call out is Ahmet Balkan's kubectx and its brother-in-arms, kubens. Both are available from the same [GitHub repo](https://github.com/ahmetb/kubectx). They give you a shorter way to switch between contexts and namespaces than their somewhat lengthy kubectl counterparts. Plus, they basically behave like the cd command for changing directories (e.g. `kubectx -` gets you back to the previous context). Nothing magic, but neat. Check the embedded videos in the GitHub readme, and you'll quickly get the idea. 

> Another trick many Kubernetes people use is to display current cluster and namespace your command line prompt. You might think this is unnecessary, but wait until you accidentally delete a pod from the wrong namespace. If you are using bash or zsh (the default shell on MacOS), you can simply use [https://github.com/jonmosco/kube-ps1](https://github.com/jonmosco/kube-ps1). You can install it via your favorite package manager, and you'll have to add a few lines to your .bashrc or .zshrc. The repository's readme describes the process well. If you're using [fish](https://fishshell.com), you may want to take a look at [https://github.com/aluxian/fish-kube-prompt] for the same purpose (tip: have a look at [fisher](https://github.com/jorgebucaran/fisher) to manage your fish extensions). You can easily turn both prompt mods on and off with `kubeon|kubeoff` for kube-ps1 and `kube_ps on|off` for fish-kube-prompt. 

Reviewer: Yeah, I've gotten way to close to doing that with our own prod... Oops

> # Handling multiple (local) Kubernetes clusters

Reviewer 2: ```suggestion
# Handling Multiple (Local) Kubernetes Clusters
```

> When you start working with Kubernetes, you will sooner or later end up with multiple k8s clusters that you need to interact with. Promise. The good news is, kubectl, the Kubernetes Command Line Interface (CLI) client, has built-in support for multiple clusters. There is however a couple of things you should know, a number of traps that you might want to avoid, and a number of tools you will want to adopt early on in your Kubernetes journey. Investing time to wrap your head around these things will save you a lot of time and hassle. 

Reviewer 2: ```suggestion
When you start working with Kubernetes, sooner or later you'll end up with multiple clusters that you need to interact with. Promise. The good news is, kubectl, the Kubernetes Command Line Interface (CLI) client, has built-in support for multiple clusters. There are, however, a couple of things you should know, a number of traps to avoid and several tools you'll want to adopt early on in your Kubernetes journey. Investing time to wrap your head around these things will save you a lot of time and hassle. 
```

> Now, this is certainly not a new topic, and all of this info can be found somewhere else. But it is spread out in a number of places, which makes it a bit challenging to figure things out - particularly when you're new to the Kubernetes world. So collecting (hopefully) all relevant information in one place and putting it into a logical order seemed useful.  

Reviewer 2: ```suggestion
Now, this is certainly not a new topic, and all of this info can be found somewhere else. But it is spread out in a number of places, which makes it challenging to figure things out - particularly when you're new to Kubernetes. So collecting (hopefully) all relevant information in one place and putting it into a logical order seems useful.  
```

> ## Kubernetes configuration files and context switching

Reviewer 2: ```suggestion
## Kubernetes Configuration Files and Context Switching
```

> Before we start discussing multiple clusters, let's take a brief look at how kubectl finds out how to talk to a particular Kubernetes cluster. By the way, if you're looking for a general introduction to kubectl, have a look at [this tutorial](https://rancher.com/learning-paths/how-to-manage-kubernetes-with-kubectl/) by Kelly Griffin. 

Reviewer 2: ```suggestion
Before we discuss multiple clusters, let's take a brief look at how kubectl finds out how to talk to a particular Kubernetes cluster. By the way, if you're looking for a general introduction to kubectl, have a look at [this tutorial](https://rancher.com/learning-paths/how-to-manage-kubernetes-with-kubectl/) by Kelly Griffin. 
```

> The most widely used mechanism to find cluster access information are so-called _kubeconfig files_. By default, kubectl will use information from the file `$HOME/.kube/config`. It is possible to point to other config files by setting the `KUBECONFIG` environment variable or by using the `--kubeconfig` flag. There are three main components in a kubeconfig file:

Reviewer 2: ```suggestion
The most widely used mechanism to find cluster access information are so-called _kubeconfig files_. By default, kubectl uses information from the file `$HOME/.kube/config`. It is possible to point to other config files by setting the `KUBECONFIG` environment variable or by using the `--kubeconfig` flag. There are three main components in a kubeconfig file:
```

* clusters: describes your - you guessed it - clusters, i.e. specifies name, API server URL and certificate authority. 
* users: the user credentials needed to authenticate to your cluster(s), i.e. name and either key pairs or username/password pairs
* context: a combination of cluster, user and namespace to exactly define the target for your subsequent kubectl operations. Even if you only have one cluster, you can either have multiple users or different namespaces to work with, so the context construct makes it easier to switch between your most frequently used combinations. 

> You can of course just manually edit your kubeconfig files, but you can also manipulate them with a number of `kubectl config` subcommands:

Reviewer 2: ```suggestion
You can, of course, manually edit your kubeconfig files, but you can also manipulate them with a number of `kubectl config` subcommands:
```

* `kubectl config --kubeconfig=config-demo set-cluster` to define clusters
* `kubectl config --kubeconfig=config-demo set-credentials` for setting user credentials
* `kubectl config --kubeconfig=config-demo set-context` to define or change contexts

> Once you have set up your clusters, users and contexts, you can easily switch between contexts using the `kubectl config use-context` command. The currently active context can be checked with `kubectl config current-context`. 

Reviewer 2: ```suggestion
Once you have set up your clusters, users and contexts, you can easily switch between contexts using the `kubectl config use-context` command. You can check the currently active context with `kubectl config current-context`. 
```

> A more detailed description of all this can be found at [https://kubernetes.io/docs/tasks/access-application-cluster/configure-access-multiple-clusters/]. 

Reviewer 2: ```suggestion
You'll find a more-detailed description of all this can be found here [https://kubernetes.io/docs/tasks/access-application-cluster/configure-access-multiple-clusters/]. 
```

> Theoretically, this could be the end of the article, if it wasn't for the fact that once the number of clusters you interact with grows beyond low single-digit numbers, keeping all information in one config file becomes, let's say, unpleasant. 

Reviewer 2: ```suggestion
Theoretically, I could end the article here. However, once the number of clusters you interact with grows beyond low single-digit numbers, keeping all information in one config file becomes, let's say, unpleasant. 
```

> ## Multiple config files

Reviewer 2: ```suggestion
## Multiple Config Files
```

> A better way (in my humble opinion) is to keep information about different clusters in different config files. There's basically two ways to handle multiple config files: the simple way and the clever way. The simple way is you manually specify the config you want to use each time you call kubectl via the `--kubeconfig` flag. The clever way is to concatenate multiple kubeconfigs into your `KUBECONFIG` environment variable like 

Reviewer 2: ```suggestion
A better way (in my humble opinion) is to keep information about different clusters in different config files. There are basically two ways to handle multiple config files: the simple way and the clever way. The simple way is by manually specifying the config you want to use each time you call kubectl via the `--kubeconfig` flag. The clever way is to concatenate multiple kubeconfigs into your `KUBECONFIG` environment variable like 
```

> A better way (in my humble opinion) is to keep information about different clusters in different config files. There's basically two ways to handle multiple config files: the simple way and the clever way. The simple way is you manually specify the config you want to use each time you call kubectl via the `--kubeconfig` flag. The clever way is to concatenate multiple kubeconfigs into your `KUBECONFIG` environment variable like 

Reviewer 2: A better way to do what?

    export KUBECONFIG=$HOME/.kube/config:second-config-file:third-config-file. 

> On Linux and Mac, the list is delimited by a colon, while for Windows you'll use a semicolon. 

Reviewer 2: ```suggestion
On Linux and Mac, the list is delimited by a colon, while for Windows, you'll use a semicolon. 
```

> The rules according to which kubectl identifies its valid configuration data from this list of files are described in detail in the [Kubernetes documentation](https://kubernetes.io/docs/concepts/configuration/organize-cluster-access-kubeconfig/), but here's the 10000 m summary:

Reviewer 2: ```suggestion
The [Kubernetes documentation](https://kubernetes.io/docs/concepts/configuration/organize-cluster-access-kubeconfig/) describes the rules according to which kubectl identifies its valid configuration data from this list of files, but here's the 10,000 meter summary:
```

* Files specified via --kubeconfig always win, no matter what's in your KUBECONFIG 
* The content of all configs specified in $KUBECONFIG is merged
* The first file to set a value or map key wins
* Never change a value or map key. Even if a later occurrence has non-conflicting values, discard that info
* The first context found in the chain wins - if you have a context with the same name in a later file, kubectl will never find it. 

> TODO: check if this is complete

Reviewer 2: Don't forget to do your TODO

You can use `kubectl config view --merge` to check what this leads to on your system. 

In case you ended up with multiple configs in one file, and you want to separate them, you can extract all information for one specific context with

    kubectl config view --minify --flatten --context=my_context > myconfig.yaml. 

## Auto-populate $KUBECONFIG

how to set up $KUBECONFIG so that it automatically picks up all files from ~/.kube/

> If the set of clusters you interact with changes often, one elegant way to deal with that is to populate your KUBECONFIG variable automatically. There's many ways to do this, but here's what I'm using. I have this line in my environment:

Reviewer 2: ```suggestion
If the set of clusters you interact with changes often, one elegant way to deal with that is to populate your KUBECONFIG variable automatically. There are many ways to do this, but here's what I'm using. I have this line in my environment:
```

    set -xg KUBECONFIG $HOME/.kube/config:(ls -m $HOME/.kube/*.yaml | tr ',' ':' | tr -d "[:blank:]") 

>  Note that I'm using fish as my default shell, hence I'm using fish's native `set` command with the `-x` flag instead of `export`. In a normal bash or zsh environment you would use `export KUBECONFIG=...` for the same purpose. 

Reviewer 2: ```suggestion
 Note that I'm using fish as my default shell; hence, I'm using fish's native `set` command with the `-x` flag instead of `export`. In a normal bash or zsh environment, you would use `export KUBECONFIG=...` for the same purpose. 
```

> This picks up `~/.kube/config as well as all files with a .yaml file extension from the ~/.kube directory, and stores them as colon separated list in my KUBECONFIG variable. So if I add a new cluster, I just have to put the config there and re-source my environment. 

Reviewer 2: ```suggestion
This picks up `~/.kube/config as well as all files with a .yaml file extension from the ~/.kube directory and stores them as a colon-separated list in my KUBECONFIG variable. So if I add a new cluster, I just have to put the config there and re-source my environment. 
```

## Useful tools and tricks

> So far we've obviously only covered Kubernetes' built-in functionality. Which is basically fine, but the are a couple of little helpers and additional tricks that are worth knowing about. 

Reviewer 2: ```suggestion
So far we've obviously only covered Kubernetes' built-in functionality. I'd like to share a couple of little helpers and additional tricks that you should know about. 
```

Author: Reworked this a bit in an upcoming commit.

> The first tool that I want to call out is Ahmet Balkan's kubectx and its brother-in-arms, kubens. Both are available from the same [GitHub repo](https://github.com/ahmetb). There's really not much to explain about them - they give you a shorter way to switch between contexts and namespaces than their somewhat lengthy kubectl counterparts, plus they basically behave like the cd command for changing directories (e.g. `kubectx -` gets you back to the previous context). Nothing magic, but neat. Check the embedded videos in the GitHub readme, and you'll get the idea. 

Reviewer 2: ```suggestion
The first tool is Ahmet Balkan's kubectx and its brother-in-arms, kubens. Both are available from the same [GitHub repo](https://github.com/ahmetb). There's not much to explain: they give you a shorter way to switch between contexts and namespaces than their somewhat lengthy kubectl counterparts. Plus, they basically behave like the cd command for changing directories (e.g. `kubectx -` gets you back to the previous context). Nothing magic, but neat. Check the embedded videos in the GitHub readme to get the idea. 
```

Author: Reworked the paragraph in my next commit.

> Another trick many Kubernetes people use is to display current cluster and namespace your command line prompt. You might think this is unnecessary, but wait until you accidentally delete a pod from the wrong namespace. If you are using bash or zsh (the default shell on MacOS), you can simply use [https://github.com/jonmosco/kube-ps1](https://github.com/jonmosco/kube-ps1). You can install it via your favorite package manager, and you'll have to add a few lines to your .bashrc or .zshrc. The process is described well in the repository's readme. If you're using [fish](https://fishshell.com), you may want to take a look at [https://github.com/aluxian/fish-kube-prompt] for the same purpose (tip: have a look at [fisher](https://github.com/jorgebucaran/fisher) to manage your fish extensions). Both prompt mods can be turned on and off easily with `kubeon|kubeoff` for kube-ps1 and `kube_ps on|off` for fish-kube-prompt. 

Reviewer 2: ```suggestion
Another trick many Kubernetes people use is to display current cluster and namespace your command line prompt. You might think this is unnecessary, but wait until you accidentally delete a pod from the wrong namespace. If you are using bash or zsh (the default shell on MacOS), you can simply use [https://github.com/jonmosco/kube-ps1](https://github.com/jonmosco/kube-ps1). You can install it via your favorite package manager -- you'll have to add a few lines to your .bashrc or .zshrc. The repository's readme describes the process well. If you're using [fish](https://fishshell.com), take a look at [https://github.com/aluxian/fish-kube-prompt] for the same purpose (tip: have a look at [fisher](https://github.com/jorgebucaran/fisher) to manage your fish extensions). You can easily turn both prompt mods on and off with `kubeon|kubeoff` for kube-ps1 and `kube_ps on|off` for fish-kube-prompt. 
```

Author: Took this on board in my incoming commit.

> When you're multitasking between multiple development projects, a handy way to point to your current cluster based on the directory you're in is a tool called `direnv`, see [https://direnv.net]. 

Reviewer 2: ```suggestion
When you're multitasking between multiple development projects, a tool called `direnv` is a handy way to point to your current cluster based on the directory you're in ( see [https://direnv.net]). 
```

> What's your favorite tool or trick for handling multiple k8s clusters? Hop over to the [SUSE & Rancher Community](https://community.suse.com) and join the [discussion](https://community.suse.com/link-to-crosspost)!

Reviewer 2: ```suggestion
What's your favorite tool or trick for handling multiple K8s clusters? Hop over to the [SUSE & Rancher Community](https://community.suse.com) and join the [discussion](https://community.suse.com/link-to-crosspost)!
```
